Excise Random from the system image

vchuravy opened the pull request:
I am not really a fan of this solution,
but it does work. Preferably this is something one could
express to the method-table so that instead of doing this at init time
in Random, we do it on at top-level and package loading does the deletion
and addition of methods at the same time.

For Random the current solution might be fine, but for LinearAlgebra we do
need a better mechanism.

vchuravy committed:
Excise Random from the system image

rfourquet commented:
What a beautiful hack! 

> For Random the current solution might be fine, but for LinearAlgebra we do need a better mechanism.

Why wouldn't that work well for LinearAlgebra?

rfourquet commented on base/Base.jl:
```diff
@@ -621,8 +621,29 @@ end
+    RANDOM_MODULE_REF = Ref{Module}()
+    # functions defined in Random
+    function rand(args...)
+        if !isassigned(RANDOM_MODULE_REF)
+            RANDOM_MODULE_REF[] = require(Random_PkgID)
+        end
```
I'm not clear on why `RANDOM_MODULE_REF` is useful. It looks like once this ref is assigned, this `rand` method doesn't exist anymore so no more call to it can check whether the ref is assigned. So the only scenario I see where this check returns "assigned" is when a parallel call to `rand` happens before `Random` is loaded, but by the time the check is done, `Random` finished loading. As the check is the first expression of the function, it seems incredibly unlikely, and it's not too harmful in this case to call again `require`... Am i missing something?

vchuravy replied:
Methods exists in a specific world, so despite us deleting it when Random is being loaded this is still callable from a previous world.

vchuravy commented:
> Why wouldn't that work well for LinearAlgebra?

We currently need to run `delete_method` as part of the init call, and can't use it during precompilation. So this may lead to unnecessary method invalidation. For `rand` the impact might not be too bad, but I am worried about `*`.

Ideally we could express method deletion to precompilation and perform it along side method insertion and not invalidate the method table after loading the package once again.

Quoting @vtjnash 

> There needs to be a sequenced-before event for the deletion

But I don't think I have the bandwidth to implement that.

Keno commented:
I happened to discuss this today with @gbaraldi @staticfloat and a few others. The big problems with this approach is that this isn't really compatible with pre-compilation. We discussed whether it would be possible to have a special method marker that causes the runtime to load a pkgimage when something looks at the method (either inference or the runtime). If we pre-declare all methods that the pkgimage is adding to base and use this mechanism for them, I don't think we actually need to bump the world age, so this should be compatible with pre-compilation. Of course, we'd also need a mechanism to prevent the pkgimage from actually adding any methods to base that weren't predeclared, but that seems easy to do.

vchuravy commented:
Agreed, when I discussed this with @vtjnash he seemed to think that it would suffice to have a method deletion marker, such that the deletion happens before we insert the new methods in the MT.

In was also thinking fo a special kind of method that knows to load a different library, but that seems to be a bigger ask. (E.g. what do we do when `@stdlib` is not on the load path)

I would definitely appreciate some help here with the runtime changes, in particular we need to solve this for Linear algebra where the numbers of methods are.many more and the cost of invalidation is much higher.

vtjnash commented:
> it would be possible to have a special method marker that causes the runtime to load a pkgimage when something looks at the method

I don't really see how that it is different from having the package actually just loaded. It is not like the methods somehow actively do things when you aren't using them, they are just special markers with data that define what things happen when someone actively intentionally looks at them.

That is a bit different from the LinAlg situation perhaps, where we almost want an LBT-like approach (lazy blas trampoline) so that you can start with just `*`, and then later define what it means for matmul at runtime, opaque to the compiler and swappable live.

vchuravy commented:
```
help?> randn
search: randn rand tand range nand rounding

  No documentation found for public symbol.

  Base.randn is a Function.

  # 1 method for generic function "randn" from Base:
   [1] randn(args...)
       @ Base.Stubs.Random stubs.jl:20
```

Need to fix that, probably by docs noticing and calling `Base.Stubs.Random.delay_initialize`

vchuravy commented on base/stubs.jl:
```diff
@@ -0,0 +1,44 @@
+
+    import Base: rand, randn
+    function rand(args...)
+        Base.invoke_in_world(delay_initialize(), rand, args...)
```
There are two stages.

1. `Random` is not loaded yet. Code is executing in world `N`. On a call to `rand`/`randn` we delay load Random into world `N+1`.
2. Code now executes in world `N+1`l `Base.Stubs.randn` no longer exists, code that used it got invalidated and everything in that uses `randn` will directly call the method in Random.

Code executing in scenario 1 has a performance penalty, which can be avoided by doing `import Random`. It's not perfect, but I believe it's the only workable compromise, that doesn't actively
break scripts. Note that in the repl only the first call to `rand` has this penalty and the next time you will execute in `N+1`

vchuravy commented:
@timholy Test failures here are due to `test/prcompile.jl` using `rand()` and the assumptions not holding anymore. Should I replace `rand()` with some other effectful function?

@KristofferC https://github.com/JuliaLang/Pkg.jl/blob/b02fb95979c71dc5834aad739ad61622cccf4a16/test/api.jl#L274 assumes that Random is in the system image? I assume deleting that test should be fine?

vchuravy commented:
@nanosoldier `runtests()`

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/51432#issuecomment-1758460553) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/7e227cb_vs_f90fd72/report.html) is available.

KristofferC commented:
>  I assume deleting that test should be fine?

Yes, that seems fine.

vchuravy committed:
Support deleting methods during precompilation for stdlib excision (#51641)

The problem with the `delete_method` in `__init__` approach is that we
invalidate the method-table,
after we have performed all of the caching work. A package dependent on
`Random`, will still see
the stub method in Base and thus when we delete the stub, we may
invalidate useful work.

Instead we delete the methods when Random is being loaded, thus a
dependent package only ever sees
the method table with all the methods in Random, and non of the stubs
methods.

The only invalidation that thus may happen are calls to `rand` and
`randn` without first doing an `import Random`.

vchuravy commented:
I am inclined towards merging this next week to start shaking out issues with this on nightlies.

@vtjnash or @JeffBezanson does either of you have time for a review?

PallHaraldsson commented:
Given that this has a performance cost for rand until you do `import Random`, and then not after, is it maybe not worth it to do this? [Random doesn't need to be added to an environment, e.g. for benchmarking scripts(?)]

What's the cost of having rand (just Xoshiro) in the sysimage? randn does only have 26% overhead over rand, so probably should be kept too (I don't know about the code overhead of it, likely larger but still small).

Given that Xoshiro, the default rng is really simple and small (at least for uniform), both its state and the code, is it possible to excise Random partially from the sysimage? I.e. the legacy rng, and all that's not usable before the import, i.e. what's not strictly in Base, but feels like it.

vchuravy committed:
Give a deprecation warning
